vyper/venom/passes/sccp/eval.py

```diff
@@ -0,0 +1,129 @@
+import operator
+from typing import Callable
+
+from vyper.utils import SizeLimits, evm_div, evm_mod
+from vyper.venom.basicblock import IROperand
+
+
+def _unsigned_to_signed(value: int) -> int:
+    if value <= SizeLimits.MAX_INT256:
+        return value
+    else:
+        return value - SizeLimits.CEILING_UINT256
+
+
+def _signed_to_unsigned(value: int) -> int:
+    if value >= 0:
+        return value
+    else:
+        return value + SizeLimits.CEILING_UINT256
+
+
+def _wrap_int_binop(operation):
+    def wrapper(ops: list[IROperand]) -> int:
+        first = _unsigned_to_signed(ops[1].value)
+        second = _unsigned_to_signed(ops[0].value)
+        return _signed_to_unsigned(int(operation(first, second)))
+
+    return wrapper
+
+
+def _wrap_uint_binop(operation):
+    def wrapper(ops: list[IROperand]) -> int:
+        first = ops[1].value
+        second = ops[0].value
+        return (int(operation(first, second))) & SizeLimits.MAX_UINT256
+
+    return wrapper
+
+
+def _evm_signextend(ops: list[IROperand]) -> int:
+    value = ops[0].value
+    bits = ops[1].value
+
+    if bits > 31:
+        return value
+
+    bits = bits * 8 + 7
+    sign_bit = 1 << bits
+    if value & sign_bit:
+        value |= SizeLimits.CEILING_UINT256 - sign_bit
+    else:
+        value &= sign_bit - 1
+
+    return value
+
+
+def _evm_iszero(ops: list[IROperand]) -> int:
+    return 1 if ops[0].value == 0 else 0
+
+
+def _evm_shr(ops: list[IROperand]) -> int:
+    value = ops[0].value
+    shift_len = ops[1].value
+    if shift_len >= 256:
+        return 0
+    return (value >> shift_len) & SizeLimits.MAX_UINT256
+
+
+def _evm_shl(ops: list[IROperand]) -> int:
+    value = ops[0].value
+    shift_len = ops[1].value
+    if shift_len >= 256:
+        return 0
+    return (value << shift_len) & SizeLimits.MAX_UINT256
+
+
+def _evm_sar(ops: list[IROperand]) -> int:
+    value = _unsigned_to_signed(ops[0].value)
+    shift_len = ops[1].value
+    if shift_len >= 256:
+        return 0 if value >= 0 else (SizeLimits.CEILING_UINT256 - 1)
+    return (value << shift_len) & SizeLimits.MAX_UINT256
+
+
+def _evm_not(ops: list[IROperand]) -> int:
+    value = ops[0].value
+    return SizeLimits.MAX_UINT256 - value
```

Comment:
i don't like this, can we do xor?
```suggestion
    return SizeLimits.MAX_UINT256 ^ value
```